New DataPointFilter : Voxel HashMap

nicolaslauzon opened the pull request:
# Description

### Summary:
This filter is inspired by [kiss-icp](https://github.com/PRBonn/kiss-icp/blob/main/cpp/kiss_icp/core/VoxelHashMap.cpp). It utilizes an HashMap to store the points in a voxel pattern. We limit the insertion to a maximum number of points per voxel which then allow for some filtering. Thanks to the HashMap, this filter is much faster than the VoxelGridDataPointFilter and is around two times slower than the RandomDownsamplingDataPointFilter.

### Filter in action:
<img width="1033" height="760" alt="image" src="https://github.com/user-attachments/assets/007a62e1-cea8-4d73-b9a3-c3a2a0932ce6" />
The original point cloud is colored in blue. The subsampled point cloud is represented by the black points.


### Changes and type of changes (quick overview):

- Feature : New DataPointFilter named VoxelHashMap
- Feature : Python binding for this new filter
- Feature : Unit test to validate the new filter
- Docs : Update the python bindings compilation doc
- Docs : Documented the new VoxelHashMapDataPointFilter

---

RedLeader962 committed:
fix: ensure compatibility with Ubuntu 20.04 in dependency installation script

- Added specific handling for Python and CMake installation on Ubuntu 20.04 systems.
- Corrected print_msg calls to use the n2st:: namespace consistently.

Youtrack issue NMO-606

RedLeader962 committed:
docs: reorganize and improve README installation instructions

- Moved quick start dependencies/install details under a new "Custom install" section.
- Added clarification and refined installer script usage with examples and options.
- Removed redundant quick start overview to streamline content.

RedLeader962 committed:
docs: fix teamcity badge

nicolaslauzon committed:
Initial implementation of the voxel hashmap data point filter

nicolaslauzon committed:
Fixed a small bug in voxelHashMap

nicolaslauzon committed:
Voxel hash map python binding

nicolaslauzon committed:
Added a test for the VoxelHashMapDataPointFilter

nicolaslauzon committed:
revert to develop readme

nicolaslauzon committed:
Revert unwanted changes

nicolaslauzon committed:
Revert unwanted changes

Revert more unwanted changes

revert unwanted changes

nicolaslauzon committed:
Revert more unwanted changes

nicolaslauzon committed:
Merge branch 'feat/voxelHashMapDataPointFilter' of github.com:norlab-ulaval/libpointmatcher into feat/voxelHashMapDataPointFilter

nicolaslauzon committed:
Merge branch 'develop' into feat/voxelHashMapDataPointFilter

nicolaslauzon committed:
Added filterProfiler example to the example build

nicolaslauzon committed:
Merge branch 'feat/voxelHashMapDataPointFilter' of github.com:norlab-ulaval/libpointmatcher into feat/voxelHashMapDataPointFilter

nicolaslauzon committed:
Update docs

boxanm commented on doc/DataFilters.md:
Mention in the docs which points are kept (first in the array / random...)

boxanm commented on pointmatcher/DataPointsFiltersImpl.h:
Can you revert the auto-formatting of this file?

boxanm commented on pointmatcher/DataPointsFilters/VoxelHashMap.cpp:
From the code, I'm not sure the points are chosen randomly as mentioned in the header file

boxanm reviewed (commented):
Good job! I added some comments asking mostly for clarification in the docs.

nicolaslauzon committed:
Clarified overall doc of the VoxelHashMapDataPointFilter and reverted autoformating on a file

nicolaslauzon commented:
> Good job! I added some comments asking mostly for clarification in the docs.

@boxanm thanks for the comments! I clarified the doc and reverted the autoformatting.
